File: day06/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <sstream>
#include  <regex>
#include <optional>
#include <functional>
#include <set>
#include <unordered_map>
#include <map>

std::vector<int64_t> find_integers(const std::string &str) {
    static std::regex base_regex("(\\d+)");

    std::vector<int64_t> result;
    for (auto it = std::sregex_iterator(str.begin(), str.end(), base_regex);
         it != std::sregex_iterator(); ++it) {

        auto value = std::stoll(it->operator[](1).str());
        result.push_back(value);
    }
    return result;
}

void part1() {

    std::fstream ifs(std::string(SRC_DIR) + "/input.txt");
    std::string line;

    std::getline(ifs, line);
    auto times = find_integers(line);
    std::getline(ifs, line);
    auto distances = find_integers(line);

    int64_t result = 1;
    for (int idx = 0; idx < times.size(); ++idx) {

        const auto time = times[idx];
        int64_t count = 0;
        for (int64_t i = 0; i <= time; ++i) {
            if (i * (time - i) > distances[idx]) {
                ++count;
            }
        }
        result *= count;
    }
    std::cout << "part1: " << result << std::endl;
}


void part2() {

    std::fstream ifs(std::string(SRC_DIR) + "/input.txt");
    std::string line;

    std::getline(ifs, line);
    line.erase(std::remove(line.begin(), line.end(), ' '), line.end());
    auto times = find_integers(line);
    std::getline(ifs, line);
    line.erase(std::remove(line.begin(), line.end(), ' '), line.end());
    auto distances = find_integers(line);


    const auto time = times[0];
    const auto distance = distances[0];
    int64_t count = 0;
    for (int64_t i = 0; i <= time; ++i) {
        if (i * (time - i) > distance) {
            ++count;
        }
    }

    std::cout << "part2: " << count << std::endl;
}


int main() {

    part1();
    part2();
    return 0;
}

File: day06/src/main.rs
use std::collections::HashSet;
use lazy_static::lazy_static;
use regex::{Match, Regex};
use std::fs::File;
use std::io::{self, BufRead};
use std::iter::Map;
use std::path::Path;
use std::slice::Iter;
use std::vec;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let file_path = "input.txt";
    let Ok(lines) = read_lines(file_path) else {
        return vec!();
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

fn find_integers(input: &str) -> Vec<i64> {

    lazy_static! {
        static ref NUMBER_RE: Regex = Regex::new(r#"(?P<number>[0-9]+)"#).unwrap();
    }

    let mut result: Vec<i64> = vec![];
    for capture in NUMBER_RE.captures_iter(input) {
        let number = capture.name("number").unwrap().as_str().parse::<i64>().unwrap();
        result.push(number);
    }
    result
}

fn part1() {

    let lines = read();
    let times = find_integers(&lines[0]);
    let distances = find_integers(&lines[1]);

    let mut result = 1i64;
    for (idx, &time) in times.iter().enumerate() {

        let mut count = 0i64;
        for i in 1..time {
            if i * (time - i) > distances[idx] {
                count+=1;
            }
        }
        result *= count;
    }

    println!("part1 {}", result);
}

fn part2() {

    let lines = read();
    let times = find_integers(&lines[0].replace(" ", ""));
    let distances = find_integers(&lines[1].replace(" ", ""));

    let mut result = 1i64;
    for (idx, &time) in times.iter().enumerate() {

        let mut count = 0i64;
        for i in 1..time {
            if i * (time - i) > distances[idx] {
                count+=1;
            }
        }
        result *= count;
    }

    println!("part2 {}", result);
}

fn main() {
    part1();
    part2();
}
